### DISCUSSION TITLE
Running GUI

### DISCUSSION THREAD

**Discussion by librarianmage** [mxgmn/MarkovJunior#40] [Q&A] [answered]
Hello,

I can't seem to figure out how to run the GUI. The code seems to be included in the source, but I haven't a clue on how to get it to start.

**Comment by Magnogen** [comment]
I've tried running the .exe in the release too, but it doesn't seem to do anything. Could it perhaps need the commandline?

**Comment by inkydragon** [comment, marked as answer]
This may be a naming confusion.

The `GUI` parameter is actually used to specify the width of the sidebar for rules in the output image.


Just modify your `models.xml`:

- `<model name="Circuit" size="100" steps="1200"/>`
![image](https://user-images.githubusercontent.com/5158738/171793237-beeac397-dc57-436c-83b6-70fc9d8bded1.png)


- `<model name="Circuit" size="100" steps="1200" gui="200"/>`
![image](https://user-images.githubusercontent.com/5158738/171792998-364d5e63-afc7-4aec-8886-44582e7182fb.png)


----

## Value of `GUI` and output image width

```

  │   GUI   │  size * pixelsize=4   │
  │◄───────►│◄─────────────────────►│
  │         │                       │

  ┌─────────┬───────────────────────┐
  │ Sidebar │                       │
  │ for     │                       │
  │ rules   │                       │
  │         │                       │
  │         │                       │
  │         │        Output         │
  │         │        image          │
  │         │                       │
  │         │                       │
  │         │                       │
  │         │                       │
  │         │                       │
  └─────────┴───────────────────────┘

```

https://github.com/mxgmn/MarkovJunior/blob/67c446e2ab64288e8a170a669e18ad9aec54b067/source/Graphics.cs#L39

eg: `<model name="Circuit" size="100" steps="1200" gui="200"/>`

For output img:
- `width = GUI + size*pixelsize = 200 + 100*4 = 600`
- `hight = size*pixelsize = 100*4 = 400`


In fact `models.xml` supports specifying the height and width of images directly, see the code for more available parameters:

https://github.com/mxgmn/MarkovJunior/blob/67c446e2ab64288e8a170a669e18ad9aec54b067/source/Program.cs#L23-L28

**Comment by nonchip** [comment]
> This may be a naming confusion.

agreed, if it's not an interface, it can't be a user interface. i'd call that more of a "state graph" or something.

**Comment by Magnogen** [comment]
Ahhh, alright. Thank you for the clarification

**Comment by gabrielelibardi** [comment]
Hi, I am running MJ on Ubuntu from the console with `dotnet run --configuration Release MarkovJunior.csproj`. It looks like it is running fine after a fix from [here](https://github.com/mxgmn/MarkovJunior/pull/4) but it is not showing anything. My goal would be to see the program as it is being executed as in the gifs. I also tried to replace the line `<model name="Circuit" size="100" steps="1200"/>` to no avail. Is there anything really obvious I am missing? I am new to C#

**Comment by mxgmn** [comment]
@gabrielelibardi add `gif="True"` to this line:
```xml
<model name="Circuit" size="100" steps="1200" gui="300" gif="True"/>
```

**Comment by gabrielelibardi** [comment]
That works thanks!

**Comment by tigrouind** [comment]
So it's not a real-time app. It actually produces a gif

**Comment by mxgmn** [comment]
It produces animation frames that you can then convert into a gif.

**Comment by AlfredBerg** [comment]
I have been playing around a bit in a fork with rendering the pixels of the gifs with sfml in a window, this way there is no need to convert all the images to a gif every time a model has been modified.
Did this mostly just to make it easier for me to play around with Markov Junior and the code is very quickly hacked together and not intended to be merged, but thought I would share it in case somebody else finds it useful.
It can be found here: https://github.com/AlfredBerg/MarkovJunior
(it is not as laggy as the gif makes it look like)
![2022-06-25 22-12-25(1)](https://user-images.githubusercontent.com/18570335/175789328-26aec6f1-bcb3-4339-b7d6-df4f526a5029.gif)
